File: Assignment1/version2/Java/Set.java
package individual.v2;

public class Set {
	
	Node head;
	Node end;
	int size;
	
	public Set(){
		head = null;
		end = null;
		size = 0;
	}
	
	//Take input and create a new set
	public Set(String input){
		//Create a array from input with comma as split regex
		String[] inputAsArray = input.split(",");
		//Cretae the head node of the set from the array's first value
		head = new Node (inputAsArray[0], null, null);
		Node current = head;
		Node prev = null;
		size = 0;
		boolean unique = true;
		
		//Go through array and add unique values to the set
		for (int i = 1; i < inputAsArray.length; i++){
			unique = true;
			//For each index of the array check if value already exist in set
			while(current != null){
				if (current.getValue().equals(inputAsArray[i])){
					unique = false;
					break;
				}
				else{
					prev = current;
					current = current.getNext();
				}
			}
			//If value does not already exist in set create a new node and add to set
			if (unique){
				Node newNode = new Node(inputAsArray[i], null, prev);
				prev.setNext(newNode);
				end = newNode;
				size++;
			}
			
			current = head;
		}
	}
	
	//Method to add new value to set
	public void add(String input){
		
		//In the event the set is empty the new value will become the head
		if (size == 0){
			Node newNode = new Node(input, null, null);
			head = newNode;
			end = newNode;
			size++;
		}
		
		//Else add new node to end of current set
		else{
			Node newNode = new Node(input, null, end);
			end.setNext(newNode);
			end = newNode;
			size++;	
		}
	}
	
	//Method to find union of two sets
	public Set union(Set set2){
		Set unionSet = new Set();
		
		Node current = this.head;
		
		//First add all values from this set
		while(current != null){
			unionSet.add(current.getValue());
			current = current.getNext();
		}
		
		current = set2.head;
		Node current2 = unionSet.head;
		boolean unique = true;
		
		//Second add values from second set if value found is unique
		while(current != null){
			unique = true;
			current2 = unionSet.head;
			while(current2 != null){
				if (current.getValue().equals(current2.getValue())){
					unique = false;
					break;
				}
				else
					current2 = current2.getNext();
			}
			if (unique){
				unionSet.add(current.getValue());	
			}
			current = current.getNext();
		}
		
		return unionSet;
	}
	
	//Method to find intersection of two set
	public Set intersection(Set set2){
		Set intersectionSet = new Set();
		
		Node current = this.head;
		Node current2 = set2.head;
		
		//Loop through first set
		while (current != null){
			current2 = set2.head;
			//Loop through second set
			//If value from second set matches value from first set
			//Add value to intersection set
			while(current2 != null){
				if (current.getValue().equals(current2.getValue())){
					intersectionSet.add(current.getValue());
					break;
				}
				else
					current2 = current2.getNext();
			}
			current = current.getNext();
		}
		
		return intersectionSet;
	}
	
	//Method to find difference of two set
	public Set difference(Set set2){
		//First find the union and intersection of two sets
		Set differenceSet = new Set();
		Set unionSet = this.union(set2);
		Set intersectionSet = this.intersection(set2);
		
		Node current = unionSet.head;
		Node current2 = intersectionSet.head;
		boolean unique = true;
		
		//Loop through union set
		while (current != null){
			unique = true;
			current2 = intersectionSet.head;
			//Add value to difference set if value from intersection set
			//Is not found in union set
			while(current2 != null){
				if (current.getValue().equals(current2.getValue())){
					unique = false;
					break;
				}
				else
					current2 = current2.getNext();
			}
			
			if (unique){
				differenceSet.add(current.getValue());
			}
			
			current = current.getNext();
		}
		
		return differenceSet;
	}
	
	//Method to print set
	public void printSet(){
		Node current = head;
		String prefix = "";
		//Loop through set starting from head and print value held in current node
		while (current != null){
			System.out.print(prefix + current.getValue());
			prefix = ",";
			current = current.getNext();
		}
		System.out.println();
	}
}

File: CS408 Project1/version2/C++/Set.cpp
#include <iostream>
#include "Set.h"
#include <string>
#include <sstream>
#include <vector>
#include <algorithm>

using namespace std;

//Constructor to create empty set
Set::Set() {
	head = NULL;
	end = NULL;
	size = 0;
}

//Constructor for set when input is given
Set::Set(string input) {
	vector<string> inputAsVector;
	head = NULL;
	end = NULL;

	stringstream ss(input);
	string value;

	//Create a vector object using user input seperated by comma
	while (getline(ss, value, ',')) {
		inputAsVector.push_back(value);
	}

	//sort and remove duplicate elements
	sort(inputAsVector.begin(), inputAsVector.end());
	inputAsVector.erase(unique(inputAsVector.begin(), inputAsVector.end()), inputAsVector.end());

	//Create new node and add them to set
	for (int i = 0; i < inputAsVector.size(); i++) {
		nodePtr newNode = new node;
		newNode->next = NULL;
		newNode->value = inputAsVector.at(i);
		
		//if head is not null just add to end of set
		//else new node becomes head
		if (head != NULL) {
			end->next = newNode;
			end = newNode;
		}
		else {
			head = newNode;
			end = newNode;
		}
		size++;
	}
}

//Method to add new node to set
void Set::addNode(nodePtr passedNode) {

	//Create a new node and pass the value from passed node
	nodePtr newNode = new node;
	newNode->value = passedNode->value;
	newNode->next = NULL;

	//if head is not null just add to end of set
	//else new node becomes head
	if (head != NULL) {
		end->next = newNode;
		end = newNode;
	}
	else {
		head = newNode;
		end = newNode;
	}
	size++;
}

//Method to add new node to set
void Set::addNode(string input) {

	//Create a new node and pass the value from passed node
	nodePtr newNode = new node;
	newNode->value = input;
	newNode->next = NULL;

	//if head is not null just add to end of set
	//else new node becomes head
	if (head != NULL) {
		end->next = newNode;
		end = newNode;
	}
	else {
		head = newNode;
		end = newNode;
	}
	size++;
}

//Method to find union of two sets
Set Set::operator + (Set &set2) {

	Set unionSet;
	nodePtr current = head;


	//Add nodes in set 1 to unionSet
	while (current != NULL) {
		unionSet.addNode(current);
		current = current->next;
	}

	nodePtr currentUnionSet = unionSet.head;
	nodePtr set2Current = set2.head;
	bool unique = true;


	//Add nodes in set 2 to unionSet if it does not exist already
	while (set2Current != NULL) {
		unique = true;
		currentUnionSet = unionSet.head;
		while (currentUnionSet != NULL) {
			if (currentUnionSet->value == set2Current->value) {
				unique = false;
				break;
			}
			else
				currentUnionSet = currentUnionSet->next;
		}
		if (unique) {
			unionSet.addNode(set2Current);
		}
		set2Current = set2Current->next;
	}

	return unionSet;
}

//Method to find intersection of two sets
Set Set::operator ^ (Set &set2) {
	Set intersectionSet;
	nodePtr set1Current = head;
	nodePtr set2Current = set2.head;

	//Loop through first and second set
	//If a element is found in both set add to intersectionSet
	while (set1Current != NULL) {
		set2Current = set2.head;
		while (set2Current != NULL) {
			if (set1Current->value == set2Current->value) {
				intersectionSet.addNode(set1Current);
				break;
			}
			else
				set2Current = set2Current->next;
		}
		set1Current = set1Current->next;
	}

	return intersectionSet;
}

//Method for difference
Set Set::operator - (Set &set2) {
	//Get the union and intersection set
	Set unionSet = (*this) + (set2);
	Set intersectionSet = (*this) ^ (set2);
	Set differenceSet;

	nodePtr unionSetCurrent = unionSet.head;
	nodePtr intersectionSetCurrent = intersectionSet.head;
	bool unique = true;

	//Loop through first and second set
	//If a element is not found in intersection set but found in union set
	//Add to difference set
	while (unionSetCurrent != NULL) {
		unique = true;
		intersectionSetCurrent = intersectionSet.head;
		while (intersectionSetCurrent != NULL) {
			if (!unionSetCurrent->value.compare(intersectionSetCurrent->value)) {
				unique = false;
				break;
			}
			else
				intersectionSetCurrent = intersectionSetCurrent->next;
		}

		if (unique) {
			differenceSet.addNode(unionSetCurrent);
		}

		unionSetCurrent = unionSetCurrent->next;
	}

	return differenceSet;
}

//Method to get the head of the set
Set::nodePtr Set::getHead() {
	return head;
}

//Method to get size of set
int Set::getSize() {
	return size;
}

//Method to print values in set
void Set::printSet() {
	nodePtr current = head;

	//Loop from head and print value held in each node
	while (current != NULL) {
		cout << current->value + " ";
		current = current->next;
	}
	cout << endl;
}